From b6f768ba8b66f4f11a555638cf84bbc790ca565c Mon Sep 17 00:00:00 2001
From: shxrrydw <sherrydrv@gmail.com>
Date: Sun, 26 Apr 2026 23:18:43 +0200
Subject: [PATCH] HACK: winex11: Skip rendering some overlay windows.

Fixes issues with a few overlay windows staying over interaction buttons in
a few game launchers depending on the compositor (e.g. NTE).

App: Neverness to Everness (4508340).
---
 dlls/winex11.drv/window.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index c2d231e71bf..da55394ba0e 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1803,6 +1803,17 @@ static int skip_iconify(void)
     return cached;
 }
 
+static BOOL skip_map_window( HWND hwnd )
+{
+    WCHAR class_name[64];
+    UNICODE_STRING class_str = { .Buffer = class_name, .MaximumLength = sizeof(class_name) };
+
+    NtUserGetClassName( hwnd, FALSE, &class_str );
+    return !wcscmp( class_name, u"CoverWindow" ) || !wcscmp( class_name, u"mainBg.xml" )
+        || !wcscmp( class_name, u"CustomInstallWndBg.xml" ) || !wcscmp( class_name, u"MessageBoxBg.xml" );
+
+}
+
 static void window_set_wm_state( struct x11drv_win_data *data, UINT new_state, BOOL activate )
 {
     UINT old_state = data->pending_state.wm_state;
@@ -1915,7 +1926,10 @@ static void window_set_wm_state( struct x11drv_win_data *data, UINT new_state, B
     case MAKELONG(WithdrawnState, NormalState):
     case MAKELONG(IconicState, NormalState):
         if (data->embedded) set_xembed_flags( data, XEMBED_MAPPED );
-        else XMapWindow( data->display, data->whole_window );
+        else if (skip_map_window( data->hwnd ))
+            WARN( "HACK: not mapping overlay window.\n ");
+        else
+            XMapWindow( data->display, data->whole_window);
         break;
     case MAKELONG(NormalState, WithdrawnState):
     case MAKELONG(IconicState, WithdrawnState):
